waspc/src/Wasp/Generator/DbGenerator/Operations.hs

```diff
@@ -20,13 +28,25 @@ printJobMsgsUntilExitReceived chan = do
     J.JobOutput {} -> printJobMessage jobMsg >> printJobMsgsUntilExitReceived chan
     J.JobExit {} -> return ()
 
-migrateDev :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-migrateDev projectDir = do
+-- | Migrates in the generated project context and then copies the migrations dir back
+-- up to the Wasp source dir to ensure they remain in sync.
+migrateDevAndSync :: Path' Abs (Dir WaspProjectDir) -> Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
+migrateDevAndSync waspDir projectDir = do
   chan <- newChan
   (_, dbExitCode) <-
     concurrently
       (printJobMsgsUntilExitReceived chan)
       (DbJobs.migrateDev projectDir chan)
   case dbExitCode of
-    ExitSuccess -> return (Right ())
+    ExitSuccess -> copyMigrationsUp waspDir projectDir
     ExitFailure code -> return $ Left $ "Migrate (dev) failed with exit code: " ++ show code
+
+copyMigrationsUp :: Path' Abs (Dir WaspProjectDir) -> Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
+copyMigrationsUp waspDir projectRootDir =
+  do
+    copyDirectoryRecursive projectMigrationsDir waspMigrationsDir >> return (Right ())
+      `catch` (\e -> return $ Left $ show (e :: P.PathException))
+      `catch` (\e -> return $ Left $ show (e :: IOError))
+  where
+    waspMigrationsDir = SP.castDir $ waspDir SP.</> dbMigrationsDirInWaspProjectDir
```

Comment:
Hm I just realized now that we have dbMigrationsDirInWaspProjectDir as a hardcoded thing in Wasp.Common, but then on the other hand we also have getMigrationsDir in Wasp (our IR). So we have it both as a hardcoded thing, and as a variable -> why is that? This might be mistake I made in the past, maybe I didn't realize there is hardcoded thing and I introduced a variable.

We should look into having just one I guess? Btw now AppSpec will replace Wasp but again, I put `migrationsDir` field into it because I saw it in Wasp and assumed it is needed, similar as we have for ext/ dir and .env file and so on.

Ok, so obviously one part of our system does need to know where migrations dir is in the wasp project (wasp project being the source, not the generated project). The question is, I guess, how far does that knowledge extend. Is it just CLI? Or is it non-Generator part of Lib (e.g. Wasp.Common), and then CLI and Generator use it? Or is it non-Generator part of Lib (e.g. Wasp.Common), and CLI uses it, but Generator receives it during runtime via Wasp IR, as getMigrationsDir field? The last sentence is how it works now, except for this copyMigrationsUp functions, which is not a generator really though (although it is a part of the Generator), instead it is an operation, which are IO actions in the context of Generator.

From what I see, so far the only place where `dbMigrationsDirInWaspProjectDir` from Wasp.Common is used is in the Lib.hs file. Which means it is very localized and not used in the Generator (until now). So maybe the last option that I described sounds good, and we could limit it to just Lib and not let Generator use it like this.

So, we could possibly say that `migrateDevAndSync` receives not `waspDir`, but instead `dbMigrationsDirInWaspProjectDirAbs :: Path' Abs (Dir Wasp.Common.DbMigrationsDir)`! That way we avoid usage Generator directly using this hardcoded value from Wasp.Common.

Reply:
So looking back into it, it seems both of those notions arose in "Propagates deletion of DB migrations dir from Wasp src to the generated project" https://github.com/wasp-lang/wasp/commit/4156cebb20cb43fc8897e7806ba615ab765eb7c1 

I originally added `dbMigrationsDirInWaspProjectDir` because we were using it in CLI Migrate and Lib (to `setMigrationsDir`). Previously, CLI Migrate had its own local definition of `[reldir|migrations|]` and we had the need in Wasp lib we wanted to keep DRY. That reference in CLI Migrate being removed in this commit. However, I think the idea of having a centrally defined notion of that dir is probably still ok for Wasp.Common and use in Lib and possibly again in CLI in future.

The `migrationsDir` was added to Wasp in this same commit since we needed that path in DBGenerator to potentially remove it.

Overall, I think you are right it would be nice if we had a clean line between their uses. In `copyMigrationsUp` we don't have access to a Wasp instance, since we are getting here via CLI. I can try to update it such that the CLI constructs a `dbMigrationsDirInWaspProjectDirAbs :: Path' Abs (Dir Wasp.Common.DbMigrationsDir)` and passes that instead of an `waspDir :: Path' Abs (Dir ProjectRootDir)` to both `migrateDevAndSync` and `copyMigrationsUp` 👍🏻 I'll give it a go, thanks

Reply:
Updated like so in commit https://github.com/wasp-lang/wasp/pull/420/commits/42a03534386d45f19aed9889088fb3f05ccf9f85:
```haskell
-- | Migrates in the generated project context and then copies the migrations dir back
-- up to the wasp project dir to ensure they remain in sync.
migrateDevAndCopyToSource :: Path' Abs (Dir DbMigrationsDir) -> Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
migrateDevAndCopyToSource dbMigrationsDirInWaspProjectDirAbs genProjectRootDirAbs = do
```